pkg/flavor/flavor.go

```diff
@@ -43,95 +44,116 @@ func NewMethods(flavorStore, clusterFlavorStore cache.Store) Methods {
+	}
+	return clusterFlavor.Spec, nil
 }
 
-func getKey(namespace string, name string) string {
-	return namespace + "/" + name
+func (m *methods) FindFlavor(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachineFlavor, error) {
+
+	if vm.Spec.Flavor == nil {
+		return nil, nil
+	}
+
+	key := vm.Namespace + "/" + vm.Spec.Flavor.Name
+	obj, exists, err := m.flavorStore.GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.NewNotFound(flavorv1alpha1.Resource(apiflavor.SingularResourceName), key)
+	}
+	flavor := obj.(*flavorv1alpha1.VirtualMachineFlavor)
+	return flavor, nil
 }
 
-func getProfiles(name string, namespace string, kind string, flavorStore, clusterFlavorStore cache.Store) ([]flavorv1alpha1.VirtualMachineFlavorProfile, error) {
-	switch strings.ToLower(kind) {
-	case apiflavor.PluralResourceName, apiflavor.SingularResourceName:
-		key := getKey(namespace, name)
-		obj, exists, err := flavorStore.GetByKey(key)
-		if err != nil {
-			return nil, err
-		}
-		if !exists {
-			return nil, errors.NewNotFound(flavorv1alpha1.Resource(apiflavor.SingularResourceName), key)
-		}
-		flavor := obj.(*flavorv1alpha1.VirtualMachineFlavor)
-		return flavor.Profiles, nil
+func (m *methods) FindClusterFlavor(vm *virtv1.VirtualMachine) (*flavorv1alpha1.VirtualMachineClusterFlavor, error) {
 
-	case "", apiflavor.ClusterPluralResourceName, apiflavor.ClusterSingularResourceName:
-		obj, exists, err := clusterFlavorStore.GetByKey(name)
-		if err != nil {
-			return nil, err
-		}
-		if !exists {
-			return nil, errors.NewNotFound(flavorv1alpha1.Resource(apiflavor.ClusterSingularResourceName), name)
-		}
-		flavor := obj.(*flavorv1alpha1.VirtualMachineClusterFlavor)
-		return flavor.Profiles, nil
-	default:
-		return nil, fmt.Errorf("got unexpected kind in FlavorMatcher: %s", kind)
+	if vm.Spec.Flavor == nil {
+		return nil, nil
 	}
+
+	key := vm.Spec.Flavor.Name
+	obj, exists, err := m.clusterFlavorStore.GetByKey(key)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		return nil, errors.NewNotFound(flavorv1alpha1.Resource(apiflavor.ClusterSingularResourceName), key)
+	}
+	flavor := obj.(*flavorv1alpha1.VirtualMachineClusterFlavor)
+	return flavor, nil
 }
 
-func applyCpu(field *k8sfield.Path, profile *flavorv1alpha1.VirtualMachineFlavorProfile, vmiSpec *virtv1.VirtualMachineInstanceSpec) Conflicts {
-	if profile.CPU == nil {
+func apply(field *k8sfield.Path, flavorSpec *flavorv1alpha1.VirtualMachineFlavorSpec, vm *virtv1.VirtualMachine, vmi *virtv1.VirtualMachineInstance) Conflicts {
+
+	if vmi.Annotations == nil {
+		vmi.Annotations = make(map[string]string)
+	}
+
+	switch strings.ToLower(vm.Spec.Flavor.Kind) {
+	case apiflavor.SingularResourceName, apiflavor.PluralResourceName:
+		vmi.Annotations[virtv1.FlavorAnnotation] = vm.Spec.Flavor.Name
+	case apiflavor.ClusterSingularResourceName, apiflavor.ClusterPluralResourceName, "":
+		vmi.Annotations[virtv1.ClusterFlavorAnnotation] = vm.Spec.Flavor.Name
+	}
```

Comment:
You can call function `AddFlavorNameAnnotations()` here.

Reply:
Ah bad rebase apologies, I think I've switched over to this later in the series. I'll switch this over here.